functional/connectivity-loss_test.go

```diff
@@ -0,0 +1,229 @@
+// Copyright 2016 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package functional
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/coreos/fleet/functional/platform"
+	"github.com/coreos/fleet/functional/util"
+)
+
+// Check that units states do not change on loss of connectivity to etcd.
+//
+// Note: this only tests the behaviour of the disconnected node;
+// but not the reaction of the rest of the cluster,
+// nor reconciliation after connectivity is restored.
+func TestSingleNodeConnectivityLoss(t *testing.T) {
+	cluster, err := platform.NewNspawnCluster("smoke")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer cluster.Destroy()
+
+	m0, err := cluster.CreateMember()
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = cluster.WaitForNMachines(m0, 1)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Set up some units.
+	stateMapping := map[string]struct {
+		command          []string
+		runState         string
+		systemdFileState string
+		systemdState     []string
+	}{
+		"inactive": {[]string{"submit"}, "", "", nil},
+		"loaded":   {[]string{"load", "--no-block"}, "inactive", "enabled", nil},
+		"launched": {[]string{"start", "--no-block"}, "active", "enabled", []string{"loaded", "active", "running"}},
+	}
+	createUnits := map[string][]string{}
+	expectedUnitFiles := map[string]string{}
+	expectedUnitStates := map[string]string{}
+	expectedSystemdFiles := map[string]string{}
+	expectedSystemdStates := map[string][]string{}
+	for _, service := range []string{"single", "global"} {
```

Comment:
I'm sort of borderline about having this nested loop and constructing the expected results vs just encoding them as fixtures. I appreciate the flexibility/extensibility of the current approach but it makes it harder to read and reason about the tests and expected results.